<a href="https://colab.research.google.com/github/geenordconsult/I-QUBE-PROJECT/blob/main/VIDEOGAMESALES__PREDICTIVE_REGRESSION.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [20]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [21]:
vgsales = pd.read_csv('vgsales.csv')
vgsales

Rank                                              Name Platform  \
0          1                                        Wii Sports      Wii   
1          2                                 Super Mario Bros.      NES   
2          3                                    Mario Kart Wii      Wii   
3          4                                 Wii Sports Resort      Wii   
4          5                          Pokemon Red/Pokemon Blue       GB   
...      ...                                               ...      ...   
16593  16596                Woody Woodpecker in Crazy Castle 5      GBA   
16594  16597                     Men in Black II: Alien Escape       GC   
16595  16598  SCORE International Baja 1000: The Official Game      PS2   
16596  16599                                        Know How 2       DS   
16597  16600                                  Spirits & Spells      GBA   

         Year         Genre   Publisher  NA_Sales  EU_Sales  JP_Sales  \
0      2006.0        Sports    Nintendo     41.49     29.02      3.77   
1      1985.0      Platform    Nintendo     29.08      3.58      6.81   
2      2008.0        Racing    Nintendo     15.85     12.88      3.79   
3      2009.0        Sports    Nintendo     15.75     11.01      3.28   
4      1996.0  Role-Playing    Nintendo     11.27      8.89     10.22   
...       ...           ...         ...       ...       ...       ...   
16593  2002.0      Platform       Kemco      0.01      0.00      0.00   
16594  2003.0       Shooter  Infogrames      0.01      0.00      0.00   
16595  2008.0        Racing  Activision      0.00      0.00      0.00   
16596  2010.0        Puzzle    7G//AMES      0.00      0.01      0.00   
16597  2003.0      Platform     Wanadoo      0.01      0.00      0.00   

       Other_Sales  Global_Sales  
0             8.46         82.74  
1             0.77         40.24  
2             3.31         35.82  
3             2.96         33.00  
4             1.00         31.37  
...            ...           ...  
16593         0.00          0.01  
16594         0.00          0.01  
16595         0.00          0.01  
16596         0.00          0.01  
16597         0.00          0.01  

[16598 rows x 11 columns]

In [22]:
vgsales.info
vgsales

Rank                                              Name Platform  \
0          1                                        Wii Sports      Wii   
1          2                                 Super Mario Bros.      NES   
2          3                                    Mario Kart Wii      Wii   
3          4                                 Wii Sports Resort      Wii   
4          5                          Pokemon Red/Pokemon Blue       GB   
...      ...                                               ...      ...   
16593  16596                Woody Woodpecker in Crazy Castle 5      GBA   
16594  16597                     Men in Black II: Alien Escape       GC   
16595  16598  SCORE International Baja 1000: The Official Game      PS2   
16596  16599                                        Know How 2       DS   
16597  16600                                  Spirits & Spells      GBA   

         Year         Genre   Publisher  NA_Sales  EU_Sales  JP_Sales  \
0      2006.0        Sports    Nintendo     41.49     29.02      3.77   
1      1985.0      Platform    Nintendo     29.08      3.58      6.81   
2      2008.0        Racing    Nintendo     15.85     12.88      3.79   
3      2009.0        Sports    Nintendo     15.75     11.01      3.28   
4      1996.0  Role-Playing    Nintendo     11.27      8.89     10.22   
...       ...           ...         ...       ...       ...       ...   
16593  2002.0      Platform       Kemco      0.01      0.00      0.00   
16594  2003.0       Shooter  Infogrames      0.01      0.00      0.00   
16595  2008.0        Racing  Activision      0.00      0.00      0.00   
16596  2010.0        Puzzle    7G//AMES      0.00      0.01      0.00   
16597  2003.0      Platform     Wanadoo      0.01      0.00      0.00   

       Other_Sales  Global_Sales  
0             8.46         82.74  
1             0.77         40.24  
2             3.31         35.82  
3             2.96         33.00  
4             1.00         31.37  
...            ...           ...  
16593         0.00          0.01  
16594         0.00          0.01  
16595         0.00          0.01  
16596         0.00          0.01  
16597         0.00          0.01  

[16598 rows x 11 columns]

In [23]:
vgsales.describe()

Rank          Year      NA_Sales      EU_Sales      JP_Sales  \
count  16598.000000  16327.000000  16598.000000  16598.000000  16598.000000   
mean    8300.605254   2006.406443      0.264667      0.146652      0.077782   
std     4791.853933      5.828981      0.816683      0.505351      0.309291   
min        1.000000   1980.000000      0.000000      0.000000      0.000000   
25%     4151.250000   2003.000000      0.000000      0.000000      0.000000   
50%     8300.500000   2007.000000      0.080000      0.020000      0.000000   
75%    12449.750000   2010.000000      0.240000      0.110000      0.040000   
max    16600.000000   2020.000000     41.490000     29.020000     10.220000   

        Other_Sales  Global_Sales  
count  16598.000000  16598.000000  
mean       0.048063      0.537441  
std        0.188588      1.555028  
min        0.000000      0.010000  
25%        0.000000      0.060000  
50%        0.010000      0.170000  
75%        0.040000      0.470000  
max       10.570000     82.740000

In [24]:
vgsales.shape

(16598, 11)

In [25]:
vgsales.duplicated().sum()

np.int64(0)

In [26]:
vgsales.isna().sum()

Rank              0
Name              0
Platform          0
Year            271
Genre             0
Publisher        58
NA_Sales          0
EU_Sales          0
JP_Sales          0
Other_Sales       0
Global_Sales      0
dtype: int64

In [27]:
vgsales.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
16593    False
16594    False
16595    False
16596    False
16597    False
Length: 16598, dtype: bool

In [28]:
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(strategy='mean')
vgsales['Year'] = imputer.fit_transform(vgsales['Year'].values.reshape(-1, 1))
vgsales.isna().sum()

Rank             0
Name             0
Platform         0
Year             0
Genre            0
Publisher       58
NA_Sales         0
EU_Sales         0
JP_Sales         0
Other_Sales      0
Global_Sales     0
dtype: int64

DATA IMPUTATION

In [29]:
imputer = SimpleImputer(strategy='most_frequent')
vgsales[['Publisher']] = imputer.fit_transform(vgsales[['Publisher']])
vgsales.isna().sum()

Rank            0
Name            0
Platform        0
Year            0
Genre           0
Publisher       0
NA_Sales        0
EU_Sales        0
JP_Sales        0
Other_Sales     0
Global_Sales    0
dtype: int64

CONVERTING CATEGORICAL DATA INTO NUMERICAL DATA

In [30]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
vgsales['Platform'] = le.fit_transform(vgsales['Platform'])
vgsales['Genre'] = le.fit_transform(vgsales['Genre'])
vgsales['Publisher'] = le.fit_transform(vgsales['Publisher'])
vgsales['Name'] =le.fit_transform(vgsales['Name'])
vgsales

Rank   Name  Platform    Year  Genre  Publisher  NA_Sales  EU_Sales  \
0          1  11007        26  2006.0     10        359     41.49     29.02   
1          2   9327        11  1985.0      4        359     29.08      3.58   
2          3   5573        26  2008.0      6        359     15.85     12.88   
3          4  11009        26  2009.0     10        359     15.75     11.01   
4          5   7346         5  1996.0      7        359     11.27      8.89   
...      ...    ...       ...     ...    ...        ...       ...       ...   
16593  16596  11101         6  2002.0      4        269      0.01      0.00   
16594  16597   5796         7  2003.0      8        241      0.01      0.00   
16595  16598   8144        16  2008.0      6         21      0.00      0.00   
16596  16599   5014         4  2010.0      5          8      0.00      0.01   
16597  16600   8967         6  2003.0      4        546      0.01      0.00   

       JP_Sales  Other_Sales  Global_Sales  
0          3.77         8.46         82.74  
1          6.81         0.77         40.24  
2          3.79         3.31         35.82  
3          3.28         2.96         33.00  
4         10.22         1.00         31.37  
...         ...          ...           ...  
16593      0.00         0.00          0.01  
16594      0.00         0.00          0.01  
16595      0.00         0.00          0.01  
16596      0.00         0.00          0.01  
16597      0.00         0.00          0.01  

[16598 rows x 11 columns]

In [31]:

X = vgsales.drop('Global_Sales', axis=1)
Y = vgsales['Global_Sales']



FEATURE SCALING

In [32]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X = sc.fit_transform(X)
Y = sc.fit_transform(Y.values.reshape(-1, 1))

In [33]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

1. Predictive Regression

Goal: Predict Global_Sales (continuous value)

Features: Platform, Year, Genre, Publisher, NA/EU/JP/Other sales

Target: Global_Sales

📌 Possible models:

Linear Regression (baseline)

Random Forest Regressor

XGBoost / CatBoost (handle categorical features well)

Neural Network (if you want to experiment)

💡 Example Question: “Given a game’s platform, genre, publisher, and regional sales, can we predict its global sales?”

🔹 2. Classification (Hit vs. Flop)

Goal: Predict whether a game is a hit or flop

Define a threshold: e.g., Global_Sales >= 1 million = hit, else flop.

📌 Models:

Logistic Regression

Decision Tree / Random Forest

Gradient Boosted Trees

💡 Example Question: “Can we classify whether a game will be a hit or not based on its features?”

🔹 3. Multi-class Classification

Goal: Predict the Genre or Platform of a game given its other attributes.

📌 Models:

Random Forest Classifier

CatBoost / XGBoost

KNN classifier (baseline)

💡 Example Question: “Given regional sales and publisher, can we predict the genre of the game?”

🔹 4. Clustering (Unsupervised)

Goal: Group similar games together.

📌 Models:

K-Means clustering (on sales data)

Hierarchical clustering

t-SNE / PCA for visualization

💡 Example Question: “Which games form natural clusters by region (NA/EU/JP sales patterns)?”

🔹 5. Time-Series / Trend Analysis

The dataset has Year → you can analyze trends:

Predict number of games released in future years (but dataset stops around 2016, so limited).

Analyze rise/fall of platforms (e.g., Wii peak years).

In [43]:
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
!pip install xgboost
from xgboost import XGBRegressor
!pip install catboost
from catboost import CatBoostRegressor




INSTANTIATING THE MODEL

In [45]:
models = {
    "Linear Regression": LinearRegression(),
    "Random Forest": RandomForestRegressor(random_state=42),
    "XGBoost": XGBRegressor(random_state=42, verbosity=0),
    "CatBoost": CatBoostRegressor(verbose=0, random_state=42)
}


PERFORMANCE METRICS

In [51]:

from sklearn.metrics import mean_squared_error, r2_score
results = []
for name, model in models.items():
    model.fit(X_train, Y_train)
    y_pred = model.predict(X_test)

    mse = mean_squared_error(Y_test, y_pred)
    rmse = mse ** 0.5
    r2 = r2_score(Y_test, y_pred)
    results.append({"Model": name, "RMSE": rmse, "R2 Score": r2, "MSE": mse})

/usr/local/lib/python3.12/dist-packages/sklearn/base.py:1389: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


In [52]:
results

[{'Model': 'Linear Regression',
  'RMSE': 0.003358867585471007,
  'R2 Score': 0.9999935069958018,
  'MSE': 1.1281991456727833e-05},
 {'Model': 'Random Forest',
  'RMSE': 0.5286358669308159,
  'R2 Score': 0.8391677384483557,
  'MSE': 0.2794558798056953},
 {'Model': 'XGBoost',
  'RMSE': 0.599513663136726,
  'R2 Score': 0.7931487795129408,
  'MSE': 0.35941663228761583},
 {'Model': 'CatBoost',
  'RMSE': 0.5748171688446121,
  'R2 Score': 0.8098399075796687,
  'MSE': 0.33041477759853527}]

1. Linear Regression

RMSE = 0.0033 → On average, your predictions are off by only 0.003 sales units (almost nothing!).

MSE = 1.12e-05 → Errors are extremely tiny.

R² = 0.99999 → The model explains ~99.999% of the variance in your data.

⚠️ Interpretation: This performance is too good to be true. Likely, Linear Regression is overfitting because the data is simple, collinear, or your features leak information about the target. It’s suspiciously perfect.

2. Random Forest

RMSE = 0.5286 → Predictions are off by about 0.53 sales units on average.

MSE = 0.2794 → Moderate squared errors.

R² = 0.839 → Explains about 84% of the variance in global sales.

✅ More realistic performance. Shows that it generalizes decently but isn’t perfect.

3. XGBoost

RMSE = 0.5995 → Errors are ~0.60 sales units.

MSE = 0.3594.

R² = 0.793 → Explains ~79% of the variance.

⚠️ Slightly worse than Random Forest. Could be due to needing better hyperparameter tuning.